sklearn/partial_dependence.py

```diff
@@ -0,0 +1,648 @@
+                               num=grid_resolution, endpoint=True)
+        axes.append(axis)
+
+    return cartesian(axes), axes
+
+
+def _exact_partial_dependence(est, target_variables, grid, X, output=None):
+    """Calculate the partial dependence of ``target_variables``.
+
+    The function will be calculated by calling the ``predict_proba`` method of
+    ``est`` for classification or ``predict`` for regression on ``X`` for every
+    point in the grid.
+
+    Parameters
+    ----------
+    est : BaseEstimator
+        A fitted classification or regression model.
+    target_variables : array-like, dtype=int
+        The target features for which the partial dependency should be
+        computed (size should be smaller than 3 for visual renderings).
+    grid : array-like, shape=(n_points, len(target_variables))
+        The grid of ``target_variables`` values for which the
+        partial dependency should be evaluated (either ``grid`` or ``X``
+        must be specified).
+    X : array-like, shape=(n_samples, n_features)
+        The data on which ``est`` was trained.
+    output : int, optional (default=None)
+        The output index to use for multi-output estimators.
+
+    Returns
+    -------
+    pdp : array, shape=(n_classes, n_points)
+        The partial dependence function evaluated on the ``grid``.
+        For regression and binary classification ``n_classes==1``.
+    """
+    n_samples = X.shape[0]
+    pdp = []
+    for row in range(grid.shape[0]):
+        X_eval = X.copy()
+        for i, variable in enumerate(target_variables):
+            X_eval[:, variable] = np.repeat(grid[row, i], n_samples)
+        if est._estimator_type == 'regressor':
+            try:
+                pdp_row = est.predict(X_eval)
+            except NotFittedError:
+                raise ValueError('Call %s.fit before partial_dependence' %
+                                 est.__class__.__name__)
+            if pdp_row.ndim != 1 and pdp_row.shape[1] != 1:
+                # Multi-output
+                if not 0 <= output < pdp_row.shape[1]:
+                    raise ValueError('Valid output must be specified for '
+                                     'multi-output models.')
+                pdp_row = pdp_row[:, output]
+            pdp.append(np.mean(pdp_row))
+        elif est._estimator_type == 'classifier':
+            try:
+                pdp_row = est.predict_proba(X_eval)
+            except NotFittedError:
+                raise ValueError('Call %s.fit before partial_dependence' %
+                                 est.__class__.__name__)
+            if isinstance(pdp_row, list):
+                # Multi-output
+                if not 0 <= output < len(pdp_row):
+                    raise ValueError('Valid output must be specified for '
+                                     'multi-output models.')
+                pdp_row = pdp_row[output]
+            pdp_row = np.log(np.clip(pdp_row, 1e-16, 1))
```

Comment:
Easy enough to add, but would this make the code more convoluted than necessary with an extra if clause?... Does the non defined nature of a zero probability cause issues there? Dumb example:

    from sklearn.ensemble import RandomForestClassifier
    import numpy as np
    X = np.array([[1,1,1,0,0,0], [2,2,2,6,6,6]]).T
    y = [0,0,0,1,1,1]
    est = RandomForestClassifier()
    est.fit(X,y)
    est.predict_log_proba(X)

Outputs:

    array([[  0., -inf],
           [  0., -inf],
           [  0., -inf],
           [-inf,   0.],
           [-inf,   0.],
           [-inf,   0.]])